edx/xblock-lti-consumer
# [BD-24] [TNL-7661] [BB-3172] LTI Improvements - Use declarative grading model on XBlock launch

This PR adds support for the declarative grading model for LTI consumer.

**JIRA tickets**: 
- https://openedx.atlassian.net/browse/TNL-7661

~~**Discussions**:~~

**Dependencies**: None

~~**Screenshots**:~~

~~**Sandbox URL**:~~

**Merge deadline**: ASAP

**Testing instructions**:
1. Make sure lti_consumer installed as per [README.rst](https://github.com/edx/xblock-lti-consumer/blob/master/README.rst)
2. Setup a new LTI block with ``Scored = True`` and Set a ``weight`` value.
3. Test if the LTI launch is successful.
4. Go to admin ``/admin/lti_consumer/ltiagslineitem/`` URL and confirm that a Line Item has been generated automatically with correct LTI Configuration, resource id, label, and Maximum Score (``weight``).
5. Generate a JWT token -
    - Go to www.jwt.io
    - select RS-256
    - Fill in both the private and public keys section in the bottom right (use public & private key used during LTI block setup)
    - Set the header and payload as follows
```
# HEADER:
{
  "alg": "RS256",
  "typ": "JWT",
  "kid": "123"
}
# PAYLOAD:
{
  "sub": "1234567890",
  "name": "John Doe",
  "admin": true,
  "iat": 1516239022
}
```
6. Use JWT to get an OIDC token
```
curl --request POST \
  --url 'http://localhost:18000/api/lti_consumer/v1/token/<USAGE_KEY>' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data grant_type=client_credentials \
  --data 'scope=https://purl.imsglobal.org/spec/lti-ags/scope/lineitem.readonly https://purl.imsglobal.org/spec/lti-ags/scope/score https://purl.imsglobal.org/spec/lti-ags/scope/result.readonly' \
  --data client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer \
  --data client_assertion=<JWT_TOKEN>
```
7. Publish score
```
curl --request POST \
  --url http://localhost:18000/api/lti_consumer/v1/lti/1/lti-ags/1/scores \
  --header 'Authorization: Bearer <OIDC_TOKEN>' \
  --header 'Content-Type: application/vnd.ims.lis.v1.score+json' \
  --data '{"timestamp": "2020-09-27T18:54:36.736+00:00","scoreGiven" : 86,"scoreMaximum" : 100,"comment" : "This is exceptional work.","activityProgress" : "Completed","gradingProgress": "FullyGraded","userId" : "<EXTERNAL_ID>"}'
```
8. The published score should be visible in the Progress tab of the course.

**Author notes and concerns**:
1. The signal seems to be called from out of request scope causing ``User`` to be null when loaded by ``django-crum`` module. As per documentation [here](https://github.com/ninemoreminutes/django-crum/blob/eeb350fc963360633b70472f06ddc6201ece5387/docs/index.rst#get_current_user) it returns ``user=None`` when not in request scope. I couldn't figure out why xblock signal is outside the request scope, when some other places seems to be working fine like [here](https://github.com/edx/edx-platform/blob/9cf2f9f298e5e8be3b3abcaadaf0b7a96d0de0df/common/djangoapps/course_modes/signals.py#L65).

Anyhow, I fixed the issue by [impersonating](https://github.com/ninemoreminutes/django-crum/blob/eeb350fc963360633b70472f06ddc6201ece5387/docs/index.rst#impersonateusernone) as ``AnonymousUser`` [here](https://github.com/open-craft/xblock-lti-consumer/blob/09c1e0a310ac5a1971dc1177334086762bee5971/lti_consumer/plugin/compat.py#L46). But I'm not sure if this is the right way to do it.
 
**Reviewers**

~~**Settings**~~
